docs/spec.md

```diff
@@ -44,6 +44,56 @@ If a method called with incorrect types the behaviour is unspecified.
 Also if a method accepts a function it must only apply the function in accordance with
 the type signature (i.e. provide the correct number of arguments of the correct types).
 
+
+## Parametricity
+
+All methods' implementations should only use type information about arguments that is known from the
+method's type signature. It's not allowed to inspect arguments or values that they produce to get
+more information about their types. In other words they should be [parametrically polymorphic](https://en.wikipedia.org/wiki/Parametric_polymorphism).
+
+For example let's take a look at Functor's `map` signature:
+
+```
+map :: Functor f => Type f ~> (a → b, f a) → f b
+```
+
+There are three type variables in it: `f`, `a`, and `b`. Also we have some restrictions:
+
+  1. `Functor f` says that `f a` is a value of a Functor.
+  2. `Type f ~>` means that we're writing implementation of `map`
+     for a [type dictionary](#type) `Type f`. In this case we know what specific
+     types `Type f` works with, so we know everything about `f`.
+
+We don't have any restrictions for types `a` and `b`, so we don't know anything about them.
+And we're not allowed to inspect them.
+
+Here's an example of a correct `map` implementation that works with a simple wrapper `{x: ...}`:
+
+```js
+X.map = (f, v) => {
+  const a = v.x
+  const b = f(a)
+  return {x: b}
+}
+```
+
+And here's one that violates parametricity requirement although fits into type signature otherwise:
```

Comment:
It's hard to came up with a good example. I have the following so far, it still seems contrived, and grew too big for a spec to my taste.

---------------

Here is another example of a Maybe type that violates parametricity requirement:

```js
Maybe.Nothing = {type: 'Nothing'}
Maybe.Just = x => {type: 'Just', value: x}
Maybe.map = (f, v) => {
  if (v.type === 'Nothing') {
    return v
  }

  const a = v.value
  const b = f(a)

  if (b === undefined) {
    return Maybe.Nothing
  }

  return Maybe.Just(b)
}
```

We make implicit conversion of `undefined` to `Nothing`.
This looks convenient because we can now do safer operations with arrays for example:

```js
const first = arr => arr[0]
const inc = n => n + 1

inc(first([])) // NaN

Maybe.map(inc, Maybe.map(first, Maybe.Just([]))) // Nothing
```

But what if `undefined` is a ligitimate value for us:

```js
const first = arr => arr[0]
const isUndefined = x => x === undefined

Maybe.map(isUndefined, Maybe.map(first, Maybe.Just([undefined]))) // Nothing, but we expect Just(true) 
```